### PR TITLE
Add a web dashboard for monitoring node resource usage

### PR DESCRIPTION
**Pull request by ls-daniel**
Following on from some discussion on the Ray mailing list about the WebUI being kind of basic, here is a PR for a dashboard that monitors node usage. It looks like this:

raydash19

I believe it to be independent of the autoscaler/node provider/cluster setup but have only tested on AWS using the Ray autoscaler.

How does it work?

On each node, during `ray up`, a python script called `reporter.py` has a Reporter class in it which periodically publishes load metrics to a Redis pubsub (a JSON blob).

On the head node, during `ray up`, a python script called `dashboard.py` is run which subscribes to these messages, collates them into a dict and does a bit of processing. It runs a REST API endpoint using aiohttp. A basic vue.js frontend is provided which uses randomly generated token auth a bit like jupyter notebooks.

You open the url, do some stuff, it updates in real time.

More metrics can be added as necessary. It would be nice to see graphs, task information, a whole bunch of stuff, see https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/ray-dev/9Ofk_jr_3us.

The dashboard announces itself during `ray up`:
```
To monitor auto-scaling activity, you can run:
  ray exec .generated.yaml  'tail -n 100 -f /tmp/ray/session_*/logs/monitor*'

To open a console on the cluster:
  ray attach .generated.yaml

To ssh manually to the cluster, run:
  ssh -i redacted ubuntu@192.168.0.2

To view the dashboard, go to:
   http://192.168.0.2:8080?token=96c73892ebcaee35029d31a7be8e9ec3215a5c9c3f864e669968865d4fb6f196
```

This PR also has a few small tweaks to ensure private IPs are printed in the autoscaler rather than external if `use_internal_ips=True`.

### CODE REVIEW

**Comment by ls-daniel** [comment]
Oops!

aiohttp is py3 only (obviously, it uses asyncio), I've overlooked this because I no longer regularly use py2.
I can change this to flask or something else, consider this a WIP.

**Comment by richardliaw** [comment]
If aiohttp is only in the dashboard component, perhaps it could be OK it
was only py3 compatible? EOL is quite close anyways.

@robertnishihara

**Comment by ericl** [comment]
Ah, I think a lot of users (including me) use py2.

**Comment by jovany-wang** [comment]
I think we could support py3 only for the dashboard.

**Comment by ericl** [comment]
I don't see any reason to break Python 2; especially for something swappable like the web handler (afaik we could even go with the built in HttpServer).

**Comment by virtualluke** [comment]
https://pythonclock.org/

**Comment by jovany-wang** [comment]
@ericl 
I think, the reasons for only supporting it in py3 are:
1). as @virtualluke mentioned, py2 will not be maintained in the near future. So many py2 projects I know have migrated to py3.
2). This is an independent feature that will not affect any other old original features. Even if someone didn't install py3, it just means he/she can't use this feature.  And this feature is a additional feature rather than a core feature.
3). We have implemented some features(job submission) like this in our internal version. They used `asyncio`, `aiohttp` which are very useful packages. But they are all in py3. I think for this additional   
 dashboard or tools, we shouldn't support it in both py2 and py3. It also makes our development process more complex.

What do you think of these?

**Comment by ericl** [comment]
I'm aware of the python clock -- but are we planning on dropping python 2 support at that time? I'm fine with this if we are.

**Comment by ericl** [comment]
Here's a HttpServer port of dashboard.py: https://gist.github.com/ericl/76cb8c025788d188ca29ae831b2438b5

It should work on both Python 2 and 3.

File: python/ray/dashboard/res/main.js
Comment (ericl) [inline python/ray/dashboard/res/main.js]: This seems to be called after the then(). When that happens you end up with an exponentially increasing number of requests.

File: python/ray/dashboard/res/main.js
Comment (ericl) [inline python/ray/dashboard/res/main.js]: How about we don't auto refresh by default?

File: python/ray/node.py
Comment (ericl) [reply python/ray/node.py]: Btw I think the logs don't show up in stdout by default for ray.init(); we should do that to make the common case easier to debug.

File: python/setup.py
Comment (ericl) [inline python/setup.py]: I think we can use normal json and HTTPServer.

For psutil, it would be nice to add the dep, but afaik @robertnishihara has reported it breaks for a lot of users.

File: python/ray/reporter.py
Comment (ericl) [inline python/ray/reporter.py]: --http-port? We can do something similar to the "old" web UI, and incrementally try higher ports until a free one is found.

File: python/ray/dashboard/index.html
Comment (ericl) [inline python/ray/dashboard/index.html]: Could we bundle these? It's a bit weird to have a internet dependency for ray.

File: python/ray/autoscaler/commands.py
Comment (ericl) [inline python/ray/autoscaler/commands.py]: I don't think this should be tied with the autoscaler, rather it should be printed on ray init (similar to the "old" web UI url).

File: python/setup.py
Comment (robertnishihara) [inline python/setup.py]: We can't make `psutil` a hard requirement because installation of `psutil` is too flaky and leads to a lot of errors. I removed it in https://github.com/ray-project/ray/pull/2892 (see also https://github.com/ray-project/ray/issues/2806).

We could have a separate `pip install ray[webui]` or something like that.

File: python/setup.py
Comment (robertnishihara) [inline python/setup.py]: What are the advantages of this over `json`? We're already using `json` in a bunch of places, and it'd be great to keep the number of `json` libraries to a minimum.

File: python/ray/dashboard/dashboard.py
Comment (robertnishihara) [inline python/ray/dashboard/dashboard.py]: You can do `pubsub(ignore_subscribe_messages=True)` to avoid one of the cases below.

**Comment by robertnishihara** [comment]
@ls-daniel this looks awesome. Thanks for contributing this!

Regarding the Python 2 conversation, I think we should continue supporting Python 2 for now for two main reasons:
1. **I think the cost of doing so is very low.** It certainly hasn't been much of a burden yet. That said, if supporting Python 2 is a substantial burden for the dashboard, then we should discuss what the options are.
2. **I think there are substantial benefits to doing so.** Many people will continue to use Python 2 beyond the end of life. It's one thing for `numpy==1.17` or whatever to drop support for Python 2, because Python 2 users can continue using `numpy==1.16`, which is basically the same thing. However, Ray is rapidly improving and Python 2 users will probably benefit a lot from the changes we push after the end of life.

File: python/setup.py
Comment (robertnishihara) [reply python/setup.py]: I don't think we should get rid of `psutil` from `reporter.py`. Instead, I think we should just not start the UI if `psutil` is not installed.

File: python/setup.py
Comment (ls-daniel) [reply python/setup.py]: Breaks in what sense?

simplejson is only used to serialize psutil's namedtuples. A custom json encoder could be used instead if necessary (it surprises me that the standard json module doesn't have namedtuple_as_object).

File: python/ray/reporter.py
Comment (ls-daniel) [reply python/ray/reporter.py]: Yep.

File: python/ray/dashboard/index.html
Comment (ls-daniel) [reply python/ray/dashboard/index.html]: I had an internal discussion about this actually (my favoured approach was bundling, happy to go back to that).

File: python/setup.py
Comment (ls-daniel) [reply python/setup.py]: I'd prefer this approach. We could just dump an error message along the lines of 'not starting as psutil not installed'.

File: python/setup.py
Comment (ls-daniel) [reply python/setup.py]: simplejson has `namedtuple_as_object` which serialises psutil's namedtuples as a dict with keys->values, I'm not aware of a simple method that exists to do this in the standard module, though a custom encoder can be written, I can go ahead and do that.

**Comment by ls-daniel** [comment]
I can certainly move away from aiohttp.

In terms of actual py2 'support', beyond checking that it passes linters and so on I don't have the bandwidth to commit to it.

I don't have arguments for/against py2 support other than "I don't use it", so happy to defer to you WRT not using py3-only code.

File: python/ray/dashboard/res/main.js
Comment (ls-daniel) [reply python/ray/dashboard/res/main.js]: I think the solution to this is putting the update call in a .finally() block

File: python/ray/dashboard/res/main.js
Comment (ls-daniel) [reply python/ray/dashboard/res/main.js]: Any specific reason? Pretty much all of my use of the dashboard relies on this. A checkbox that disables it maybe?

File: python/setup.py
Comment (ls-daniel) [reply python/setup.py]: Ah, I saw this later down.
I can either re-implement this with a bunch of bash scripts or something, or just disable the dashboard if psutil is not installed. Prefer latter.

File: python/ray/dashboard/dashboard.py
Comment (ls-daniel) [reply python/ray/dashboard/dashboard.py]: Consider that the default behaviour of jupyter is to require a token because we have no idea whether users have exposed ports or not.

I'd be happy to include a --no-auth option (perhaps along with the --http-port option below) if necessary.

Replaced `secrets` with `SystemRandom` below.

**Comment by markgoodhead** [comment]
My thoughts on the py2 situation (I work with @ls-daniel):

I don't necessarily agree that the cost is low, the other side of the cost is the cost of lost contributions from other commercial developers because eventually users like us/alibaba with forks decide that it's not worth the time investment to go back and re-write a PR to remove py3 features (which adds nothing for us) vs doing new development that will end up being useful to us. It seems like @jovany-wang already has internal features they haven't contributed back to the main project and whilst I can't speak for him it sounds like the extra support requirement is a factor:

> We have implemented some features(job submission) like this in our internal version. They used `asyncio`, `aiohttp` which are very useful packages. But they are all in py3. I think for this additional
dashboard or tools, we shouldn't support it in both py2 and py3. It also makes our development process more complex.

Ultimately it adds a small extra overhead in taking internal development work and contributing it back to the main project and the sum of all of these kinds of individually small overheads (py2 support, avoiding using certain packages, fighting non-deterministic tests etc) increases the probability that useful features on other commercial user's forks don't get merged in a timely manner as those users decide to instead prioritise new development that helps them, leading to those users deciding instead to 'roll their own' from now on and hence may end up avoiding contributing back entirely (since their forks have diverged so much that the work to do so only gets harder) and instead focus on just maintaining their own fork that fits their requirements. 

Whilst I'm happy for our team to spend some of their development time doing a small amount of work to comply with main repo merge requirements in order to contribute back their work (we'd very much like to support the project with new development and contribute back) I have to balance that with the cost of doing so vs working on other useful features internally.

Perhaps a compromise would be to do as Numpy have and end support in the next version (i.e. Python 2 support up until 0.7.0) at which point that requirement can be dropped and Python 2 users have the option of using 0.7 if they wish to use ray, knowing it will be at the expense of missing out on new features/development. That seems to be the path that most major projects are going anyway and I don't see why Ray shouldn't follow that, especially as Ray is much newer than Numpy and will have much fewer (or indeed, any?) legacy users. Alternatively, if py2 support is a real priority for your team, we're happy producing PRs based on py3 code that we're happy with and then handing over to you if you'd like to re-factor it to be py2 compliant.

File: python/setup.py
Comment (ls-daniel) [reply python/setup.py]: Commit below disables the dashboard if `psutil` not present (the dashboard runs but the reporters don't send it anything, so it's just empty).

File: python/ray/dashboard/index.html
Comment (ls-daniel) [reply python/ray/dashboard/index.html]: Bundled, removed axios dep

File: python/setup.py
Comment (ls-daniel) [reply python/setup.py]: `reporter` bails out if psutil not present.

File: python/ray/dashboard/dashboard.py
Comment (ls-daniel) [reply python/ray/dashboard/dashboard.py]: Done

File: python/setup.py
Comment (ls-daniel) [reply python/setup.py]: Created a recursive method that works for this simple case.

File: python/ray/reporter.py
Comment (ls-daniel) [reply python/ray/reporter.py]: Done.

File: python/ray/autoscaler/commands.py
Comment (ls-daniel) [reply python/ray/autoscaler/commands.py]: Makes sense, will look at this tomorrow.

File: python/ray/dashboard/res/main.js
Comment (ericl) [reply python/ray/dashboard/res/main.js]: I'm thinking about developers mainly. It was harder to figure out what was happening on the web console due to the large number of requests, and also on the server.

Would be nice to be static by default, with a link to `#autoRefresh`

File: python/ray/dashboard/dashboard.py
Comment (ericl) [reply python/ray/dashboard/dashboard.py]: Two cases I'm thinking of:

Bookmarking for a dashboard, and refreshing a page after restarting a cluster. Having the URL be random would be impediment to both use cases.

Here I'm comparing to Spark UI which always has a fixed URL.

For security, I think it's sufficient to avoid binding to 0.0.0.0. Rather, by default you can only serve requests originating from the same machine (which ray port forward would also work for). That way you get network level security without any inconveniences.

**Comment by robertnishihara** [comment]
@ls-daniel @markgoodhead those are good points regarding contributions. Maybe the right approach here for certain non-core features is to accept PRs that only work in Python 3.

File: python/ray/node.py
Comment (ls-daniel) [reply python/ray/node.py]: Hopefully fixed now.

File: python/ray/autoscaler/commands.py
Comment (ls-daniel) [reply python/ray/autoscaler/commands.py]: Should be fixed now. The non-AS method prints a localhost url; the actual IP is written to a file (in house we're rsyncing this file down to get the url out).

**Comment by ls-daniel** [comment]
The issues on the thread have been resolved now, I think.

The dashboard and reporter only start if you're running py3. Their dependencies have been removed and they don't run if psutil/setproctitle/aiohttp are not present. Errors are logged.

File: python/ray/dashboard/dashboard.py
Comment (robertnishihara) [inline python/ray/dashboard/dashboard.py]: `/tmp/ray` is shared by all Ray sessions, but the dashboard is specific to a single Ray session I believe, so I moved this file into `/tmp/ray/session...`.

File: python/ray/node.py
Comment (robertnishihara) [inline python/ray/node.py]: I made this change to start the dashboard instead of the Jupyter notebook.

File: python/ray/services.py
Comment (robertnishihara) [inline python/ray/services.py]: I'm not sure why, but I can't seem to connect to the dashboard using `localhost` or `127.0.0.1`. I have to use the value returned by `ray.services.get_node_ip_address()`.

File: python/ray/services.py
Comment (robertnishihara) [inline python/ray/services.py]: `start_ray_process` doesn't raise an exception even if the process fails to start, so I moved the `try/except` earlier.

**Review by robertnishihara — COMMENTED** [review]
@ls-daniel I pushed a few changes, I hope you don't mind! Please take a look at the changes I made.

@ericl note that I reverted the bundling of `vue.js` and `vue.min.js` because these files add up to almost 10MB, which is a lot to commit in the repo. This means we'll have an internet dependency right now, but I think we can get rid of that by bundling those files with our wheels during installation (instead of committing them).

**Comment by ericl** [comment]
I'm fine with merging this without py2 support. In a followup we can always port it to normal HttpServer https://gist.github.com/ericl/76cb8c025788d188ca29ae831b2438b5 as needed.

**Comment by robertnishihara** [comment]
I had to rebase to fix `tempfile_test.py`.

**Comment by raulchen** [comment]
I'm also in favor of dropping Python 2 support. 

Regarding the cost to support Python 2, I think that's actually not trivial. @guoyuhong and I have encountered some tricky problems because of inconsistent behaviors between Python 2 and Python 3. Usually we had to invest considerable amount of time investigating some language internal details. And sometimes had to ended up with hacky workarounds. 

Besides this, another bigger issue in my opinion is that it prevents us from adopting more advanced technologies. For example, without changing ray internal functions to async (`async def`), we cannot make ray fully async (@suquark may have more insights to share regarding asyncio).

I think now we can open an issue to collect feedbacks from ray users regarding wether supporting Python 2 is still necessary. 

For this PR, I agree that we can make this feature Python 3 only.

**Comment by robertnishihara** [comment]
There are some test failures that I don't understand yet, but are probably related to this PR. I can help look into them tomorrow (haven't been able to reproduce them yet), so made need to try creating the same environment as is in Travis.

**Comment by ls-daniel** [comment]
Had a very brief look just now and I suspect the test fails are due to the services not running if optional dependencies aren't installed.

Either you'll need to skip checking those, or install the optional deps for the test suite.

File: python/ray/services.py
Comment (ls-daniel) [reply python/ray/services.py]: In `dashboard.py` we explicitly bind to that IP.

Should just bind to 0.0.0.0. Line 184.

**Review by robertnishihara — APPROVED** [review]
LGTM once tests pass.

**Comment by robertnishihara** [comment]
Thanks @ls-daniel!

**Comment by Deepak3994** [comment]
> Following on from some discussion on the Ray mailing list about the WebUI being kind of basic, here is a PR for a dashboard that monitors node usage. It looks like this:
> 
> raydash19
> 
> I believe it to be independent of the autoscaler/node provider/cluster setup but have only tested on AWS using the Ray autoscaler.
> 
> How does it work?
> 
> On each node, during `ray up`, a python script called `reporter.py` has a Reporter class in it which periodically publishes load metrics to a Redis pubsub (a JSON blob).
> 
> On the head node, during `ray up`, a python script called `dashboard.py` is run which subscribes to these messages, collates them into a dict and does a bit of processing. It runs a REST API endpoint using aiohttp. A basic vue.js frontend is provided which uses randomly generated token auth a bit like jupyter notebooks.
> 
> You open the url, do some stuff, it updates in real time.
> 
> More metrics can be added as necessary. It would be nice to see graphs, task information, a whole bunch of stuff, see https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/ray-dev/9Ofk_jr_3us.
> 
> The dashboard announces itself during `ray up`:
> 
> ```
> To monitor auto-scaling activity, you can run:
>   ray exec .generated.yaml  'tail -n 100 -f /tmp/ray/session_*/logs/monitor*'
> 
> To open a console on the cluster:
>   ray attach .generated.yaml
> 
> To ssh manually to the cluster, run:
>   ssh -i redacted ubuntu@192.168.0.2
> 
> To view the dashboard, go to:
>    http://192.168.0.2:8080?token=96c73892ebcaee35029d31a7be8e9ec3215a5c9c3f864e669968865d4fb6f196
> ```
> 
> This PR also has a few small tweaks to ensure private IPs are printed in the autoscaler rather than external if `use_internal_ips=True`.

@ls-daniel 

How did you get that cpu utilization from worker nodes. Which file you are using to get those details?

**Comment by robertnishihara** [comment]
@Deepak3994 it's in https://github.com/ray-project/ray/blob/master/python/ray/reporter.py. Cpu utilization is `psutil.cpu_percent()` I believe.

**Comment by Deepak3994** [comment]
Thanks @robertnishihara . I will look into this and revert you back.

**Pull request merged**
